### DISCUSSION TITLE
Contributor Roles and Teams

### DISCUSSION THREAD

**Discussion by gasparnagy** [reqnroll/Reqnroll#67] [Announcements]
I have setup the following teams for the Reqnroll GitHub organization:

- [contributors](https://github.com/orgs/reqnroll/teams/contributors)
  - the members have write access to the main repositories and can create branches there to make PRs easier without needing to fork (based on everyone's preference forks can still be used of course)
  - the "main" branch is protected, so they can only push to main through PRs
  - everyone with an approved PR with code change *within a year* will be part of the team (currently maintained manually)
- [core-team](https://github.com/orgs/reqnroll/teams/core-team)
  - permissions of "contributor" team, plus
  - can approve/merge PRs
  - can push to "main" branch (only in special cases, the normal workflow should use PRs)
  - members will be asked to give feedback on every strategic decisions
  - members can be notified using the `@core-team` mention form in discussions & issues
  - people with activities in bigger or multiple areas / discussions will be part of the team

Members of any org team required to have two-factor GitHub authentication.

Any feedback on this is appreciated.

**Comment by 304NotModified** [comment]
Some feedback :)

On https://github.com/reqnroll/Reqnroll/pull/479#issuecomment-2678448513:

- The [contributors team](https://github.com/orgs/reqnroll/teams/contributors/members) link  won't work unless your in a/the team  - that's a bit confusing
- The invitation was a bit hard to find, but we can't do much about that I guess

On above:
> the members have write access to the main repositories and can create branches there to make PRs easier (without needing to fork)

I think this isn't a good idea. Nowadays it's really easy to keep your fork up-to-date (there is a nice sync button). As with open source, contributors arrive and leave, and now there is the problem we get stale branches and those are hard to delete (when?). See also  https://github.com/reqnroll/Reqnroll/branches/stale and https://github.com/reqnroll/Reqnroll.VisualStudio/branches/stale

IMO this won't make it easier, but less easy as I also created a branch on the origin while otherwise the web UI makes it in my fork. 

PS:  I'm happy that I could edit/update Github issues and run the pipelines. That really helps in keeping things up-to-date and completing the PRs for review.

> can push to "main" branch and approve/merge PRs

I never like to push to main without PRs. It's tricky. I propose that all changes needs a PR - maybe that's already the case, not sure :)

**Reply by gasparnagy** [reply]
Thx for the feedback and thx for your recent active contribution!

> The [contributors team](https://github.com/orgs/reqnroll/teams/contributors/members) link won't work unless your in a/the team - that's a bit confusing

OK. Removed the link.

> The invitation was a bit hard to find, but we can't do much about that I guess

Yeah... if there is an easy way to find it, please let me know, I can include it into the message.

> I think this isn't a good idea. Nowadays it's really easy to keep your fork up-to-date (there is a nice sync button).
> IMO this won't make it easier, but less easy as I also created a branch on the origin while otherwise the web UI makes it in my fork.

I don't want to make a big deal about this. I think this is subjective (I very rarely use the UI for example). It says "can", not "must" 😉. Maybe I should emphasize it in the text that you can use forks as well. 

But it happened many times now that we helped each-other on a PR which only works conveniently if you are in the main repo. But this is obviously more important for the bigger features or for the complex changes.

I always take care of the stale branches 😎 - the ones listed there are not abandoned. And I always delete the branch once the PR is merged (can be done with a simple click). Let's monitor this, but so far this was not a problem.

> I never like to push to main without PRs. It's tricky. I propose that all changes needs a PR - maybe that's already the case, not sure :)

This is only for the core team. They use it wisely. It's a useful backup solution though. E.g. when I need to fix a typo in the changelog before doing the release I push directly to main. I will update the text above to make this less confusing.

**Reply by gasparnagy** [reply]
Updated the text accordingly.
